### DISCUSSION TITLE
integration into QML

### DISCUSSION THREAD

**Discussion by Be-ing** [slint-ui/slint#303] [Ideas] [answered]
I am really intrigued to see this in the [0.0.6 release announcement](https://sixtyfps.io/blog/v0.0.6-now-with-ide-support):

> This release also includes new Rust and C++ APIs to dynamically load .60 files, with the sixtyfps-interpreter crate in Rust, and the sixtyfps::interpreter library in C++. This also allows embedding a component as a QWidget in a Qt application.

Are there plans to make it possible to integrate SixtyFPS with QML? That would open a path to gradually migrating old C++ applications using QML to Rust. Are there technical barriers to doing this?

**Comment by tronical** [comment]
I think the biggest technical barrier is the complexity of the input. General purpose qml written by hand would be extremely difficult to reliably translate. Perhaps something could be done with tool generated qml that’s a bit more limited.

Either way I think it would be a translation that’s always affected with some loss, so if done it might be best suitable on a file/project level, to generate`.60` files that can be edited further.

You have a good point though: It makes sense to explore a migration path from C++/QML to Rust/.60. Perhaps initially as a written guide, to help migrate concepts. So also on a language level not a generic C++ to Rust guide but focused on the patterns of Qt/C++.

**Reply by Be-ing** [reply]
I think you misunderstood my question. I am asking about how to use SixtyFPS within QML or QML within SixtyFPS. I am not asking how to translate .qml files to .60 files, although that would be awesome.

**Comment by ogoffart** [comment, marked as answer]
> That would open a path to gradually migrating old C++ applications using QML to Rust.

Unrelated, but that the scope of another project of mine: https://github.com/woboq/qmetaobject-rs
Using that project, you would still use QML with Rust. So no SixtyFPS :-/

> Are there plans to make it possible to integrate SixtyFPS with QML?

We did integration with QWidget because that we needed for a specific commercial use-case.
Similarily, an integration with QML could be done. (Diplaying .60 code within a QML application, or even embedding a QML scene in a SixtyFPS application.)
This is currently not on our short term roadmap, but that could change if there is the need to do that.
